unit/analyses/ai/ai.cpp

```diff
@@ -0,0 +1,297 @@
+/*******************************************************************\
+
+ Module: Unit tests for selectively retaining particular domains in ait
+
+ Author: Diffblue Ltd.
+
+\*******************************************************************/
+
+/// \file
+/// Unit tests for selectively retaining particular domains in ait
+
+#include <testing-utils/catch.hpp>
+
+#include <analyses/ai.h>
+
+#include <ansi-c/ansi_c_language.h>
+
+#include <goto-programs/goto_convert_functions.h>
+
+#include <langapi/mode.h>
+
+#include <util/arith_tools.h>
+#include <util/config.h>
+#include <util/c_types.h>
+#include <util/optional.h>
+
+/// A very simple analysis that counts executed instructions along a particular
+/// path, taking the max at merge points and saturating at 100 instructions.
+/// It should indicate that instructions not within a loop have a certain path
+/// length, and those reachable from a loop have a path length of 100
+/// (i.e. potentially infinity)
+class instruction_counter_domaint : public ai_domain_baset
+{
+public:
+
+  optionalt<unsigned> path_length;
+
+  void transform(locationt, locationt, ai_baset &, const namespacet &) override
+  {
+    if(*path_length < 100)
+      ++*path_length;
+  }
+
+  void make_bottom() override
+  {
+    path_length = {};
+  }
+  void make_top() override
+  {
+    UNREACHABLE;
+  }
+  void make_entry() override
+  {
+    path_length = 0;
+  }
+  bool is_bottom() const override
+  {
+    return !path_length.has_value();
+  }
+  bool is_top() const override
+  {
+    UNREACHABLE;
+    return true;
+  }
+
+  bool merge(const instruction_counter_domaint &b, locationt from, locationt to)
+  {
+    if(b.is_bottom())
+      return false;
+
+    if(!path_length.has_value())
```

Comment:
This is always false as the condition `b.is_bottom()` checked before is the same.

Reply:
Not so, the test above checks the incoming domain; this checks the current domain.